server/block/campfire.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// Activate ...
+func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User, ctx *item.UseContext) bool {
+	if held, _ := u.HeldItems(); !held.Empty() {
+		if _, ok := held.Item().(item.FlintAndSteel); ok {
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			c.Extinguished = false
+			w.SetBlock(pos, c, nil)
+			return true
+		}
+		if rawFood, ok := held.Item().(item.Smeltable); ok && rawFood.SmeltInfo().Food {
+			for i, it := range c.Items {
+				if it.Item.Empty() {
+					c.Items[i] = CampfireItem{
+						Item: held.Grow(-held.Count() + 1),
+						Time: 600,
+					}
+					ctx.SubtractFromCount(1)
+					w.PlaySound(pos.Vec3Centre(), sound.AddItem{})
+					w.SetBlock(pos, c, nil)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (c Campfire) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	if _, ok := w.Block(pos).(Campfire); ok && face == cube.FaceUp {
+		return false
+	}
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return
+	}
+	c.Facing = user.Facing().Opposite()
+	place(w, pos, c, user, ctx)
+	return placed(ctx)
+}
+
+// Tick is called to cook the items within the campfire
+func (c Campfire) Tick(_ int64, pos cube.Pos, w *world.World) {
+	if !c.Extinguished {
+		if rand.Float64() <= 0.016 { // Every three or so seconds.
+			w.PlaySound(pos.Vec3Centre(), sound.CampfireCrackle{})
+		}
+		for i, it := range c.Items {
+			if !it.Item.Empty() && it.Time <= 0 {
+				itemCooked := it.Item
+				if food, ok := itemCooked.Item().(item.Smeltable); ok {
+					dropItem(w, food.SmeltInfo().Product, mgl64.Vec3{rand.Float64()*0.2 - 0.1, 0.2, rand.Float64()*0.2 - 0.1})
+					c.Items[i].Item = item.Stack{}
+					w.SetBlock(pos, c, nil)
+					continue
+				}
+			}
+			c.Items[i].Time = it.Time - 1
+			w.SetBlock(pos, c, nil)
+		}
+	}
+}
+
+// NeighbourUpdateTick ...
+func (c Campfire) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	// If the campfire is waterlogged, we need to extinguish it.
+	if _, ok := w.Liquid(pos); ok && !c.Extinguished {
+		c.Extinguished = true
+		w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+		w.SetBlock(pos, c, nil)
+	}
+}
+
+// EntityInside ...
+func (c Campfire) EntityInside(pos cube.Pos, w *world.World, e world.Entity) {
+	if flammable, ok := e.(flammableEntity); ok {
+		if flammable.OnFireDuration() > 0 && c.Extinguished {
+			c.Extinguished = false
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			w.SetBlock(pos, c, nil)
+		}
+		if !c.Extinguished && !w.RainingAt(pos) {
+			if l, ok := e.(livingEntity); ok && !l.AttackImmune() {
+				l.Hurt(c.Type.Damage(), damage.SourceFire{Campfire: true})
+			}
+			if flammable.OnFireDuration() < time.Second*8 {
+				flammable.SetOnFire(8 * time.Second)
+			}
+		}
+	}
+}
+
+// EncodeNBT ...
+func (c Campfire) EncodeNBT() map[string]any {
+	m := map[string]any{
+		"id": "Campfire",
+	}
+	for i, v := range c.Items {
```

Comment:
Might be worth adding an EncodeNBT/DecodeNBT to `CampfireItem`; see `FireworkExplosion`, for example.

Reply:
I don't believe campfire items hold any specific nbt data as items. Only when they're placed as blocks.

Reply:
![image](https://user-images.githubusercontent.com/39247615/182067946-c1c01bc2-ed54-4eeb-8c9f-b216052aaa28.png)

Reply:
That's nbt data on the block about the specific items being cooked so the client can properly display it. It isn't applicable when the campfire is an actual item.

Reply:
Don't believe it's necessary. It's boilerplate and each CampfireItem requires a number and returning map[string]any would not allow me to simply return the appropriate data. I think its fine the way it is rn.

Reply:
> CampfireItem requires a number and returning map[string]any would not allow me to simply return the appropriate data
Ah yeah that's true, ignore this then.